<h3>Data Analytics III</h3>
<hr>
<p>1. Implement Simple Naïve Bayes classification algorithm using Python on iris.csv dataset.</p>
<p>2. Compute Confusion matrix to find TP, FP, TN, FN, Accuracy, Error rate, Precision, Recall on the given dataset.</p> 
<p>
<b>Name:</b> Udayan Chavan
<b>Roll no:</b> 31117
<b>Batch:</b> K1
</p>

In [1]:
import pandas as pd
import numpy as np

import warningserror
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv("iris.csv")
data.head()

sepallength  sepalwidth  petallength  petalwidth        class
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

In [3]:
data.describe()

sepallength  sepalwidth  petallength  petalwidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.054000     3.758667    1.198667
std       0.828066    0.433594     1.764420    0.763161
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

Assigning the 4 independent variables to X and 1 dependent variable to Y:

In [22]:
X = data.iloc[:,:4].values
Y = data['class'].values

print(X[:5])
print('-'*20)
print(Y[:5])

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
--------------------
['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa']


Splitting the data into training and testing datasets:

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

Scaling data down to a smaller range:

In [11]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Training the Naïve Bayes classification model on the training dataset:

In [12]:
from sklearn.naive_bayes import GaussianNB

classifier = GaussianNB()
classifier.fit(X_train, Y_train)

GaussianNB()

Prediciting results of the test dataset:

In [29]:
Y_pred = classifier.predict(X_test)

print("Predicted data:")
print(Y_pred[:15])

print('-'*65)

print("Actual data:")
print(Y_test[:15])

Predicted data:
['Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-versicolor'
 'Iris-virginica' 'Iris-setosa' 'Iris-versicolor' 'Iris-virginica'
 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor' 'Iris-virginica'
 'Iris-setosa' 'Iris-versicolor' 'Iris-versicolor']
-----------------------------------------------------------------
Actual data:
['Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-versicolor'
 'Iris-virginica' 'Iris-setosa' 'Iris-versicolor' 'Iris-virginica'
 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor' 'Iris-virginica'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor']


As we cam see only one prediction is inccorect, hence the accuracy of our model is good. Creating a confusion matrix for further evaluation:

In [37]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)

print("Confusion Matrix:")
print(cm)

print('-'*28)

from sklearn.metrics import accuracy_score
print("Accuracy:", accuracy_score(Y_test, Y_pred))

Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  1  5]]
----------------------------
Accuracy: 0.9333333333333333


<p>For a confusion matrix made for multi-class classification:</p>
<p><b><u>Setosa class:</u></b></p>
<p>1. <b>TP:</b> (cell 1) = <b>11</b> </p>
<p>2. <b>FP:</b> (cell 4 + cell 7) = 0 + 0 = <b>0</b> </p>
<p>3. <b>TN:</b> (cell 5 + cell 6 + cell 8 + cell 9) = 12 + 1 + 1 + 5 = <b>19</b> </p>
<p>4. <b>FN:</b> (cell 2 + cell 3) = 0 + 0 = <b>0</b> </p>
------------------------------------------------------------------------------
<p> <b>Accuracy:</b> (TP + TN) / (TP + TN + FP + FN) = (11+19) / (11+19+0+0) = 30/30 = <b>1</b></p>
<p> <b>Error rate:</b> (FP + FN) / (TP + TN + FP + FN) = (0+0) / (11+19+0+0) = <b>0</b></p>
<p> <b>Precision:</b> (TP) / (TP + FP) = (11)/(11 + 0) = 11/11 = <b>1</b></p>
<p> <b>Recall:</b> (TP) / (TP + FN) = (11)/(11 + 0) = 11/11 = <b>1</b></p>

<hr>

<p><b><u>Versicolor class:</u></b></p>
<p>1. <b>TP:</b> (cell 5) = <b>12</b> </p>
<p>2. <b>FP:</b> (cell 2 + cell 8) = 0 + 1 = <b>1</b> </p>
<p>3. <b>TN:</b> (cell 1 + cell 3 + cell 7 + cell 9) = 11 + 0 + 0 + 5 = <b>16</b> </p>
<p>4. <b>FN:</b> (cell 4 + cell 6) = 0 + 1 = <b>1</b> </p>
------------------------------------------------------------------------------
<p> <b>Accuracy:</b> (TP + TN) / (TP + TN + FP + FN) = (12+16) / (12+16+1+1) = 28/28 = <b>1</b></p>
<p> <b>Error rate:</b> (FP + FN) / (TP + TN + FP + FN) = (1+1) / (12+16+1+1) = 2/30 = <b>0.067</b></p>
<p> <b>Precision:</b> (TP) / (TP + FP) = (12)/(12 + 1) = 12/13 = <b>0.92</b></p>
<p> <b>Recall:</b> (TP) / (TP + FN) = (12)/(12 + 1) = 12/13 = <b>0.92</b></p>

<hr>

<p><b><u>Viriginica class:</u></b></p>
<p>1. <b>TP:</b> (cell 9) = <b>5</b> </p>
<p>2. <b>FP:</b> (cell 3 + cell 6) = 0 + 1 = <b>1</b> </p>
<p>3. <b>TN:</b> (cell 1 + cell 2 + cell 4 + cell 5) = 11 + 0 + 0 + 12 = <b>23</b> </p>
<p>4. <b>FN:</b> (cell 7 + cell 8) = 0 + 1 = <b>1</b> </p>
------------------------------------------------------------------------------
<p> <b>Accuracy:</b> (TP + TN) / (TP + TN + FP + FN) = (5+23) / (5+23+1+1) = 30/30 = <b>1</b></p>
<p> <b>Error rate:</b> (FP + FN) / (TP + TN + FP + FN) = (1+1) / (5+23+1+1) = 2/30 = <b>0.067</b></p>
<p> <b>Precision:</b> (TP) / (TP + FP) = (5)/(5 + 1) = 5/6 = <b>0.83</b></p>
<p> <b>Recall:</b> (TP) / (TP + FN) = (5)/(5 + 1) = 5/6 = <b>0.83</b></p>
